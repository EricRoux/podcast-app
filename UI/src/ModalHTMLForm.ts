function AuthFormHTML(): string {
    return `
    <form class="mui-form auth-form">
        <div class="mui-textfield">
            <input type="email" id="email" class="authEmail" required>
            <label for="email">Email</label>
        </div>
        <div class="mui-textfield">
            <input type="password" id="password" class="authPassword" required>
            <label for="password">Пароль</label>
        </div>
        <div class="buttons">
            <div class="auth-button">
                <button 
                    type="submit"
                    id="authSubmit"
                    class="mui-btn mui-btn--raised mui-btn--primary"
                >
                    Войти
                </button>
            </div>
            <div class="auth-close-button">
                <button 
                    type="submit"
                    class="mui-btn mui-btn--raised mui-btn--primary"
                >
                    Закрыть
                </button>
            </div>
        </div>
    </form>
    `;
}

function regFormHTML(): string {
    return `
    <form class="mui-form reg-form">
        <div class="mui-textfield">
            <input type="email" id="email" class="regEmail" required>
            <label for="email">Email</label>
        </div>
        <div class="mui-textfield">
            <input type="password" id="password" class="regPassword1" required>
            <label for="password">Пароль</label>
        </div>
        <div class="mui-textfield">
            <input type="password" id="password2" class="regPassword2" required>
            <label for="password">Пароль</label>
        </div>
        <div class="buttons">
            <div class="reg-button">
                <button 
                    type="submit" 
                    id="regSubmit"
                    class="mui-btn mui-btn--raised mui-btn--primary"
                >
                    Зарегистрироваться
                </button>
            </div>
            <div class="reg-close-button">
                <button 
                    type="submit"
                    class="mui-btn mui-btn--raised mui-btn--primary"
                >
                    Закрыть
                </button>
            </div>
        </div>
    </form>
    `;
}

export function getAuthFormHTML(): string {
    return `
    <ul class="mui-tabs__bar">
        <li class="mui--is-active">
            <a data-mui-toggle="tab" data-mui-controls="pane-default-1">Авторизация</a>
        </li>
        <li>
            <a data-mui-toggle="tab" data-mui-controls="pane-default-2">Регистрация</a>
        </li>
    </ul>
    <div class="mui-tabs__pane mui--is-active" id="pane-default-1">` + AuthFormHTML() + `</div>
    <div class="mui-tabs__pane" id="pane-default-2">` + regFormHTML() + "</div>";
}