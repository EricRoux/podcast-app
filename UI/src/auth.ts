import { IAuthResponse } from "./Interfaces/IAuthResponse";
import { IUserAuth } from "./Interfaces/IUserAuth";

function AuthFormHTML(): string {
    return `
    <form class="mui-form" id="auth-form">
        <div class="mui-textfield">
            <input type="email" id="authEmail" required>
            <label for="authEmail">Email</label>
        </div>
        <div class="mui-textfield">
            <input type="password" id="authPassword" required>
            <label for="authPassword">Пароль</label>
        </div>
        <button 
            type="submit" 
            class="mui-btn mui-btn--raised mui-btn--primary"
        >
            Войти
        </button>
    </form>
    `;
}

function regFormHTML(): string {
    return `
    <form class="mui-form">
        <div class="mui-textfield">
            <input type="email" id="regEmail" required>
            <label for="regEmail">Email</label>
        </div>
        <div class="mui-textfield">
            <input type="password" id="regPassword1" required>
            <label for="password">Пароль</label>
        </div>
        <div class="mui-textfield">
            <input type="password" id="regPassword2" required>
            <label for="password">Пароль</label>
        </div>
        <button 
            type="submit" 
            class="mui-btn mui-btn--raised mui-btn--primary"
        >
            Зарегистрироваться
        </button>
    </form>
    `;
}

export function getAuthFormHTML(): string {
    return `
    <ul class="mui-tabs__bar">
        <li class="mui--is-active">
            <a data-mui-toggle="tab" data-mui-controls="pane-default-1">Авторизация</a>
        </li>
        <li>
            <a data-mui-toggle="tab" data-mui-controls="pane-default-2">Регистрация</a>
        </li>
    </ul>
    <div class="mui-tabs__pane mui--is-active" id="pane-default-1">` + AuthFormHTML() + `</div>
    <div class="mui-tabs__pane" id="pane-default-2">` + regFormHTML() + "</div>";
}

export function authWithEmailAndPassword(
    userAuth: IUserAuth
): Promise<string> {
    return fetch("http://localhost:5050/api/v1/Auth/login", {    
        method: "POST",
        body: JSON.stringify(userAuth),
        headers: {
            "Content-Type": "application/json"
        }
    })
        .then((response: Response): Promise<IAuthResponse> => response.json())
        .then((data: IAuthResponse): string => {
            if(data.status == 1)
                return data.token;
            throw new Error("Something went wrong");
        });
}