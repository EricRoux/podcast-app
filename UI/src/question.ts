import { IQuestion } from "./Interfaces/IQuestion";
import { addQuestionToLocalStorage } from "./Utils/addQuestionToLocalStorage";
import { IQuestionResponse } from "./Interfaces/IQuestionResponse";

export class Question {
    create(question: IQuestion): Promise<void> {
        return fetch("http://localhost:5050/api/v1/newQuestion", {
            method: "POST",
            body: JSON.stringify(question),
            headers: {
                "Content-Type": "application/json"
            }, 
        })
            .then((response: Response): Promise<IQuestionResponse> => {
                return response.json();
            })
            .then((json: IQuestionResponse): IQuestion => {
                question.id = json.id;
                return question;
            })
            .then(addQuestionToLocalStorage);
    }
}