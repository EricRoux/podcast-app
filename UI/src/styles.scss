/**
 * Body CSS
 */
 html,
 body {
     height: 100%;
 }
 html,
 body,
 input,
 textarea,
 button {
     -webkit-font-smoothing: antialiased;
     -moz-osx-font-smoothing: grayscale;
     text-shadow: 1px 1px 1px rgba(0, 0, 0, 0.004);
 }

 /**
  * Sidebar CSS
  */
 
 @media (min-width: 768px) {
     #sidebar {
         position: fixed;
         top: 0;
         bottom: 0;
         width: 180px;
         height: 100%;
         padding-top: 30px;
     }
 }
 /**
  * Content CSS
  */
 @media (min-width: 768px) {
     #content {
         margin-left: 180px;
     }
 }
 
.author {
    font-size: .8rem;
}

body {
    #sidebar {
        background-color: #2196f3;
        padding: 15px 25px 15px 11px;
    }
    .floating-btn {
        position: fixed;
        bottom: 50px;
        right: 50px;
        z-index: 1001;
    }
    .modal {
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        position: fixed;
        background-color: rgba(0,0,0,.2);
        overflow-y: auto;
        z-index: 1000;
        .modal-form {
            max-width: 600px;
            min-width: 265px;
            max-height: 400px;
            min-height: 200px;
            background-color: #f3ecec;
            margin: 100px auto;
            padding: auto;
            border: 1px solid #3f4d58 ;
            > h1 {
                text-align: center;
            }
            .model-content  {
                margin: 1rem;
            }
        }
    }
}
