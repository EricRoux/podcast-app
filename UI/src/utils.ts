import { IQuestion } from "./Interfaces/IQuestion";

export function isValid(value: string): boolean {
    return value.length >= 10 && value.length <= 256;
}

function getQuestionFromLocalStorage(): IQuestion[] {
    return JSON.parse(localStorage.getItem("questions") || "[]");
}

export function addQuestionToLocalStorage(question: IQuestion): void {
    const allQuestion: IQuestion[] = getQuestionFromLocalStorage();
    allQuestion.push(question);
    const allQuestionToString: string = JSON.stringify(allQuestion);
    localStorage.setItem("questions", allQuestionToString);
}