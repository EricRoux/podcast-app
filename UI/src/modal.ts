import { createModal } from "./Utils/createModal";
import { getAuthFormHTML } from "./ModalHTMLForm";
import { IUserAuth } from "./Interfaces/IUserAuth";
import { checkEmail, checkPasswordDiff, checkPasswordLen } from "./Utils/CheckCreds";
import { postRequest, URL } from "./auth";
import { createErrorMessage } from "./Utils/createErrorMessage";

export class Modal {
    private modalBtn: HTMLButtonElement;
    private modalClass: string = "modal";

    constructor(modalBtn: HTMLButtonElement) {
        this.modalBtn = modalBtn;
        this.createBtnEvents = this.createBtnEvents.bind(this);
        this.modal = this.modal.bind(this);
        this.openModal = this.openModal.bind(this);
        this.closeModal = this.closeModal.bind(this);
        this.modalEvents = this.modalEvents.bind(this);
        this.authFormHandler = this.authFormHandler.bind(this);
        this.regFormHandler = this.regFormHandler.bind(this);
        this.getCreds = this.getCreds.bind(this);
        this.closeButtonModal= this.closeButtonModal.bind(this);
    }
    
    private closeModal(): void {
        document.querySelector(`.${this.modalClass}`).remove();
        this.modalBtn.innerText = "+";
    }

    private getCreds(className: string): IUserAuth {
        return {
            email: (<HTMLInputElement>document
                .querySelector(className)
                .querySelector("#email")
            ).value,
        
            password: (<HTMLInputElement>document
                .querySelector(className)
                .querySelector("#password")
            ).value,
        };
    }
    
    private authFormHandler(event: Event): void {
        event.preventDefault();
        const userAuth: IUserAuth = this.getCreds(".auth-form");
        if(!checkEmail(userAuth.email)) {
            createErrorMessage("Неправильный Email",".auth-form");
            return;
        }
        this.closeModal();
        postRequest(URL.Login, userAuth);
    }

    private regFormHandler(event: Event): void {
        event.preventDefault();
        const userAuth: IUserAuth = this.getCreds(".reg-form");
        const password2: string = (<HTMLInputElement>document
            .querySelector(".reg-form")
            .querySelector("#password2"))
            .value;
        if(!checkEmail(userAuth.email)) {
            createErrorMessage("Неправильный Email",".reg-form");
            return;
        } else if(!checkPasswordLen(userAuth.password)){
            createErrorMessage("Пароль слишком короткий",".reg-form");
            return; 
        } else if(!checkPasswordDiff(userAuth.password, password2)) {
            createErrorMessage("Пароли не совпадают",".reg-form");
            return;
        }
        this.closeModal();
        postRequest(URL.Registration, userAuth);
    }

    private closeButtonModal(event: Event): void {
        event.preventDefault();
        this.closeModal();
    }
    
    private modalEvents(): void {
        const authForm: HTMLDivElement = document.querySelector("#authSubmit");
        authForm.addEventListener("click", this.authFormHandler);
        const authClose: HTMLDivElement = document.querySelector(".auth-close-button");
        authClose.addEventListener("click", this.closeButtonModal);
        const regForm: HTMLDivElement = document.querySelector("#regSubmit");
        regForm.addEventListener("click", this.regFormHandler);
        const regClose: HTMLDivElement = document.querySelector(".reg-close-button");
        regClose.addEventListener("click", this.closeButtonModal);        
    }

    private openModal(className: string): void {
        createModal(className, "Авторизация", getAuthFormHTML());
        this.modalEvents();
    }
    
    private modal(): void {
        if(this.modalBtn.innerText == "+"){
            this.openModal(this.modalClass);
            this.modalBtn.innerText = "-";
        } else if(this.modalBtn.innerText == "-") {
            this.closeModal();
            this.modalBtn.innerText = "+";
        } else {
            this.closeModal();
        }
    }

    createBtnEvents(): void {
        this.modalBtn.addEventListener("click", this.modal);
    }
}