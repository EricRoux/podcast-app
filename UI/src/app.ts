import "./styles.scss";
import { isValidInput } from "./Utils/isValidInput";
import { IQuestion } from "./Interfaces/IQuestion";
import { Question } from "./question";
import { renderList } from "./Utils/renderList";
import { createModal } from "./Utils/createModal";

const form: HTMLElement = document.querySelector("#form");
const input: HTMLInputElement = form.querySelector("#question-input");
const submitBtn: HTMLButtonElement = form.querySelector("#submit");
const modalBtn: HTMLButtonElement = document.body.querySelector("#modal-btn");

function submitFormHandler(event: Event): void {
    event.preventDefault(); // Не перезагружать страницу
    if (isValidInput(input.value)) {
        const questionData: IQuestion = {
            text: input.value.trim(),
            date: new Date().toJSON(),
        };

        submitBtn.disabled = true;
        const question = new Question();
        question.create(questionData)
            .then((): void => {
                input.value = "";
                input.className = "";
                renderList();
            });
    }
}

function imputChanged(): void {
    submitBtn.disabled = !isValidInput(input.value);
}

function openModal(): void {
    createModal("Авторизация", "<h1>Test</h1>");
}

window.addEventListener("load", renderList);
form.addEventListener("submit", submitFormHandler);
input.addEventListener("input", imputChanged);
modalBtn.addEventListener("click", openModal);