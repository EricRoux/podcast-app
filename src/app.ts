import "./styles.scss";
import { isValid } from "./utils";

const form: HTMLElement = document.querySelector("#form");
const input: HTMLInputElement = form.querySelector("#question-input");
const submitBtn: HTMLButtonElement = form.querySelector("#submit");

interface question {
    text: string,
    date: string,
}

function submitFormHandler(event: Event): void {
    event.preventDefault(); // Не перезагружать страницу
    if (isValid(input.value)) {
        const question: question = {
            text: input.value.trim(),
            date: new Date().toJSON(),
        };

        submitBtn.disabled = true;
        console.log("Question", question);

        input.value = "";
        input.className = "";
        submitBtn.disabled = false;
    }
}

function imputChanged(): void {
    submitBtn.disabled = !isValid(input.value);
}

form.addEventListener("submit", submitFormHandler);
input.addEventListener("input", imputChanged);