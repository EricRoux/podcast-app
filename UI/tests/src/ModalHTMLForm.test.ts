import { getAuthFormHTML } from "../../src/ModalHTMLForm";

describe("Создание Modal окна", (): void => {
    test("Получение окна", (): void => {
        expect(getAuthFormHTML.toString()).toMatchSnapshot();
    });
});