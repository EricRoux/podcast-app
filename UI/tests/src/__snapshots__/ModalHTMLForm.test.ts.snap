// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Создание Modal окна Получение окна 1`] = `
"function getAuthFormHTML() {
  /* istanbul ignore next */
  cov_qjdhe30ap().f[2]++;
  cov_qjdhe30ap().s[4]++;
  return \`
    <ul class=\\"mui-tabs__bar\\">
        <li class=\\"mui--is-active\\">
            <a data-mui-toggle=\\"tab\\" data-mui-controls=\\"pane-default-1\\">Авторизация</a>
        </li>
        <li>
            <a data-mui-toggle=\\"tab\\" data-mui-controls=\\"pane-default-2\\">Регистрация</a>
        </li>
    </ul>
    <div class=\\"mui-tabs__pane mui--is-active\\" id=\\"pane-default-1\\">\` + AuthFormHTML() + \`</div>
    <div class=\\"mui-tabs__pane\\" id=\\"pane-default-2\\">\` + regFormHTML() + \\"</div>\\";
}"
`;
